baserun/instrumentation/instrumented_wrapper.py

```diff
@@ -0,0 +1,120 @@
+import logging
+from types import GeneratorType
+from typing import Callable, TYPE_CHECKING
+
+from opentelemetry import context as context_api, trace
+from opentelemetry.context import _SUPPRESS_INSTRUMENTATION_KEY
+from opentelemetry.sdk.trace import _Span
+from opentelemetry.trace import (
+    SpanKind,
+    Status,
+    StatusCode,
+    get_current_span,
+)
+
+from baserun.instrumentation.span_attributes import SpanAttributes
+
+if TYPE_CHECKING:
+    from baserun.instrumentation.base_instrumentor import BaseInstrumentor
+
+logger = logging.getLogger(__name__)
+
+
+def instrumented_wrapper(
+    wrapped_fn: Callable, instrumentor: "BaseInstrumentor", span_name: str
+):
+    """Generates a function (`instrumented_function`) which instruments the original function (`wrapped_fn`)"""
+
+    def instrumented_function(*args, **kwargs):
+        """Replacement for the original function (`wrapped_fn`). Will perform instrumentation, call `wrapped_fn`,
+        perform more instrumentation, and then return the result of the previous call to `wrapped_fn`.
+        """
+        from baserun import Baserun
+
+        tracer_provider = trace.get_tracer_provider()
+        tracer = tracer_provider.get_tracer("baserun")
+
+        if context_api.get_value(_SUPPRESS_INSTRUMENTATION_KEY):
+            return wrapped_fn(*args, **kwargs)
+
+        parent_span: _Span = get_current_span()
+        request_type = span_name.split(".")[-1]
+        parent_span.set_attribute(SpanAttributes.LLM_REQUEST_TYPE, request_type)
```

Comment:
What's the reason for this parent span?

Reply:
We get the parent span and grab the run data from the parent span (a parent span == a run) so that we can inject the run data into the span.